gms/feature_service.cc

```diff
@@ -85,6 +85,8 @@ feature_config feature_config_from_db_config(const db::config& cfg, std::set<sst
     }
     if (!cfg.enable_tablets()) {
         fcfg._disabled_features.insert("TABLETS"s);
+    } else if (cfg.force_gossip_topology_changes()) {
+        throw std::runtime_error("Tablets cannot be enabled with gossip topology changes.  Use either --enable-tablets or --force-gossip-topology-changes, not both.");
```

Comment:
Isn't this part of raft recovery? Does this mean we cannot recover raft group 0 if tablets are enabled?

@kbr-scylla

Reply:
yes

Reply:
actually:
- force_gossip_topology_changes is not part of raft recovery
- we cannot recover raft group 0 if you have any tablet enabled keyspace

Reply:
Ok, so this patch isn't making it worse.

But we need to fix recovery. Maybe recovery mode should run without any user keyspaces enabled (and with the front-end ports closed), so there would be no tablets keyspaces.

Reply:
> Maybe recovery mode should run without any user keyspaces enabled (and with the front-end ports closed), so there would be no tablets keyspaces.

This has nothing to do with the reason why recovery mode is incompatible with tablets.

Please consider the procedure as it is described in our documentation: https://opensource.docs.scylladb.com/stable/troubleshooting/handling-node-failures.html
![Screenshot from 2024-10-16 17-55-07](https://github.com/user-attachments/assets/0849ed90-4035-413d-8ed3-182cdb75c917)

And it should become clear why it's incompatible with tablets:

It's incompatible with tablets because you're performing `nodetool removenode` using gossip-based topology as part of the manual recovery procedure (step 4). This obviously does not perform the necessary tablet migrations/rebuilding since the tablets control plane was only implemented in raft-based topology. So once you removed all the dead nodes in gossip mode, you're left with a tablet mapping that doesn't make any sense (it refers to replicas that no longer exist in the cluster etc.)

Furthermore at the end of recovery procedure you perform upgrade-to-raft-topology again (steps 9, 10), which builds raft-topology state (system.topology table and so on) from scratch based on gossiper state (system.peers table and so on) , and this logic also does not take any leftover previous tablet map into account (e.g. if system.topology referred to some tablet mappings before you started recovery, the newly built system.topology will no longer have any of this information, as there is no code implemented to build it).

Reply:
I see. Well, we need to fix it.